### PR TITLE
Adding web vault compatibility

### PR DESCRIPTION
**Pull request by Pschittt**
Update of the api to be compatible with the web vault.

What is working  : 
- Create account
- Connect
- Create/modify folders
- Create/modify ciphers
- Modify the master password (need to create a POST /password route)
- import tool

What doesn't work : 
- Managing organisation (password share)

Sorry for this ruby code, first time reading and trying to understand how it works. I'm convinced that we can optimise it.

Need to investigate the need of using cookies or not.

### CODE REVIEW

**Comment by Pschittt** [comment]
@jcs do you plan to review the code ?

**Comment by andrzejwp** [comment]
@Pschittt please review my pull-request to your patch-1 branch (https://github.com/Pschittt/bitwarden-ruby/pull/1) and potentially resubmit this pull request as your code as in commit 36e441f was throwing a syntax error.

**Comment by Pschittt** [comment]
@andrzejwp code reviewed (from my window...) and tested. Everything ok for me.
Thanks for your review.
Like I said, first time reading and writing ruby, so maybe it can be optimized.

Hope @jcs will merge into his "official" repo.

**Comment by Pschittt** [comment]
Change masterpassword available.

**Comment by andrzejwp** [comment]
@jcs I'm not fluent in Ruby either, so please review and correct whatever we did wrong here. It now works in my case (tested in Chromium and Firefox, using both the web vault as well as browser extensions), which was to create master password, log in, add and consume secrets.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+RUN addgroup -g ${GID} abc 
+RUN adduser -D -u ${UID} -G abc abc 
+RUN chown -R ${UID}:${GID} ${APP_HOME}/db
+
+EXPOSE ${PORT}
+VOLUME ["${APP_HOME}/db"]
+USER abc
+
+ENTRYPOINT ["bundle","exec"]
+CMD rackup -p ${PORT} config.ru
```
Comment (jcs) [inline Dockerfile]: Adding a Docker file should be done in a different PR.

File: lib/api.rb [hunk 2]
```diff
@@ -59,6 +59,14 @@ def validation_error(msg)
...
       params.merge!(JSON.parse(js))
     end
+  ## needed for the web vault, which doesn't use the content-type  
+  elsif request.accept.to_s.match(/application\/json/)
+    js = request.body.read.to_s
+    if !request.content_type.to_s.match(/application\/x-www-form-urlencoded/)
+      if !js.strip.blank?
+        params.merge!(JSON.parse(js))
+      end
+    end
```
Comment (jcs) [inline lib/api.rb]: This block should be merged into the one above it, to unify JSON reading and parsing (just checking `request.accept` vs `request.content_type`.

File: lib/api.rb [hunk 3]
```diff
@@ -94,9 +102,9 @@ def validation_error(msg)
       need_params(
         :client_id,
         :grant_type,
-        :deviceidentifier,
-        :devicename,
-        :devicetype,
+        #:deviceidentifier,
+        #:devicename,
+        #:devicetype,
```
Comment (jcs) [inline lib/api.rb]: Just remove these lines instead of commenting them out.

File: lib/api.rb [hunk 4]
```diff
@@ -110,11 +118,11 @@ def validation_error(msg)
 
       u = User.find_by_email(params[:username])
       if !u
-        return validation_error("Invalid username")
+        return validation_error("Invalid username or password")
       end
 
       if !u.has_password_hash?(params[:password])
-        return validation_error("Invalid password")
+        return validation_error("Invalid password or password")
```
Comment (jcs) [inline lib/api.rb]: I think you meant to have that say "Invalid username or password"

File: lib/api.rb [hunk 5]
```diff
@@ -128,13 +136,14 @@ def validation_error(msg)
...
-        # wat
-        d.destroy
-        d = nil
+      if params[:deviceidentifier]
+        d = Device.find_by_uuid(params[:deviceidentifier])
+        if d && d.user_uuid != u.uuid
+          # wat
+          d.destroy
+          d = nil
+        end
```
Comment (jcs) [inline lib/api.rb]: A previous chunk removed the enforcement of `:deviceidentifier`, `:devicename`, and `:devicetype` via `need_params()` but when `:deviceidentifier` is present (from non-web-vault requests), you need to still enforce that `:devicename` and `:devicetype` are present, and move the later assignments of them to `d.type` and `d.name` into this block after enforcement.

File: lib/api.rb [hunk 6]
```diff
@@ -157,14 +166,20 @@ def validation_error(msg)
         return validation_error("Unknown error")
       end
 
+      status 200
```
Comment (jcs) [inline lib/api.rb]: `status 200` is the default so this shouldn't be needed

File: lib/api.rb [hunk 7]
```diff
@@ -157,14 +166,20 @@ def validation_error(msg)
...
+      headers \
+      "access-control-allow-origin" => "*"
+
       {
         :access_token => d.access_token,
         :expires_in => (d.token_expires_at - Time.now).floor,
         :token_type => "Bearer",
         :refresh_token => d.refresh_token,
         :Key => d.user.key,
+	:PrivateKey => d.user.private_key,
```
Comment (jcs) [inline lib/api.rb]: Indentation

File: lib/api.rb [hunk 8]
```diff
@@ -157,14 +166,20 @@ def validation_error(msg)
...
         # TODO: when to include :privateKey and :TwoFactorToken?
       }.to_json
+
```
Comment (jcs) [inline lib/api.rb]: Please remove

File: lib/api.rb [hunk 9]
```diff
@@ -208,6 +223,8 @@ def validation_error(msg)
       u.password_hash = params[:masterpasswordhash]
       u.password_hint = params[:masterpasswordhint]
       u.key = params[:key]
+      u.public_key = params[:keys][:publicKey]
+      u.private_key = params[:keys][:encryptedPrivateKey]
```
Comment (jcs) [inline lib/api.rb]: These probably require validation.

File: lib/api.rb [hunk 10]
```diff
@@ -218,7 +235,9 @@ def validation_error(msg)
       if !u.save
         return validation_error("User save failed")
       end
-
+      status 200
+      headers \
+      "access-control-allow-origin" => "*"
```
Comment (jcs) [inline lib/api.rb]: `status 200` shouldn't be needed, and the `headers` line can be put on one line.

File: lib/api.rb [hunk 11]
```diff
@@ -157,14 +166,20 @@ def validation_error(msg)
         return validation_error("Unknown error")
       end
 
+      status 200
+      headers \
+      "access-control-allow-origin" => "*"
```
Comment (jcs) [inline lib/api.rb]: This can be on one line

File: lib/api.rb [hunk 12]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
     }.to_json
   end
 
+  # Used by the web vault to update the private and public keys if the users doesn't have ones.
```
Comment (jcs) [inline lib/api.rb]: Grammar: ... if the "user" doesn't have "one".

File: lib/api.rb [hunk 13]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+  post "/accounts/keys" do
+    content_type :json
+    # Needed by the web vault for EVERY response
+	  response['access-control-allow-origin'] = '*'
```
Comment (jcs) [inline lib/api.rb]: indentation

File: lib/api.rb [hunk 14]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+    d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+	d.user.private_key = params[:encryptedPrivateKey]
+	d.user.public_key = params[:publicKey]
```
Comment (jcs) [inline lib/api.rb]: Indentation, and these parameters need to be validated.

File: lib/api.rb [hunk 15]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+	  "Culture" => d.user.culture,
+	  "TwoFactorEnabled" => d.user.totp_secret,
+	  "Key" => d.user.key,
+	  "PrivateKey" => d.user.private_key,
+	  "SecurityStamp" => d.user.security_stamp,
+	  "Organizations" => "[]",
+	  "Object" => "profile",
+	}.to_json
+	
+
```
Comment (jcs) [inline lib/api.rb]: Please fix indentation (soft 2 spaces everywhere) and remove blank lines.

File: lib/api.rb [hunk 16]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+	  "MasterPasswordHint" => d.user.password_hint,
+	  "Culture" => d.user.culture,
+	  "TwoFactorEnabled" => d.user.totp_secret,
+	  "Key" => d.user.key,
+	  "PrivateKey" => d.user.private_key,
+	  "SecurityStamp" => d.user.security_stamp,
+	  "Organizations" => "[]",
+	  "Object" => "profile",
+	}.to_json
+  end
```
Comment (jcs) [inline lib/api.rb]: Same here

File: lib/api.rb [hunk 17]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+	rescue Bitwarden::InvalidCipherString
+	  return validation_error("Invalid key")
+	end
+	
+	if d.user.password_hash == params[:masterpasswordhash]
+      d.user.key=params[:key]
+	  d.user.password_hash=params[:newmasterpasswordhash]
+	else
+	  return validation_error("Wrong current password")
+	end
```
Comment (jcs) [inline lib/api.rb]: Please fix all of the indentation here, and spaces around `=`.

File: lib/api.rb [hunk 18]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+	  
+    User.transaction do
+	  if !d.user.save
+	    return validation_error("Unknown error")
+	  end
+	end
+	""
```
Comment (jcs) [inline lib/api.rb]: Indentation

File: lib/api.rb [hunk 19]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+  end
+
+  # Used to update email 
+  post "/accounts/email-token" do
+	content_type :json
+	response['access-control-allow-origin'] = '*'
+	validation_error("Not implemented yet")
+  end
```
Comment (jcs) [inline lib/api.rb]: Indentation

File: lib/api.rb [hunk 20]
```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
...
+	response['access-control-allow-origin'] = '*'
+
+	d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+
+	return validation_error("import tool not implemented yet")
+
+  end
```
Comment (jcs) [inline lib/api.rb]: What does this function do?

File: lib/api.rb [hunk 21]
```diff
@@ -285,6 +421,50 @@ def validation_error(msg)
...
+    Cipher.transaction do
+      if !c.save
+        return validation_error("error saving")
+      end
+
+      c.to_hash.merge({
+        "Edit" => true,
+      }).to_json
+    end
+  end
```
Comment (jcs) [inline lib/api.rb]: This does the same thing as `put "/ciphers/:uuid"` just with a different HTTP verb, so both should just call a common function.

File: lib/api.rb [hunk 22]
```diff
@@ -349,6 +529,51 @@ def validation_error(msg)
...
+	response['access-control-allow-origin'] = '*'
+	c = nil
+
+    if !(c = Cipher.find_by_uuid(params[:uuid]))
+	  return validation_error("invalid cipher")
+	end
+	c.to_hash.merge({
+        "Edit" => true,
+	}).to_json
+  end
```
Comment (jcs) [inline lib/api.rb]: Indentation and spacing for these 4 functions

File: lib/api.rb [hunk 23]
```diff
@@ -470,12 +695,22 @@ def validation_error(msg)
...
+  ### Organizations
+
+  post "/organizations" do
+    d= device_from_bearer
+	if !d
+	 return validation_error("invalid bearer")
+	end
+
+	return validation_error("Organizations not implemented yet")
+  end
```
Comment (jcs) [inline lib/api.rb]: Is this needed if it doesn't do anything yet?

File: Dockerfile (hunk 1 above)
Comment (Pschittt) [reply Dockerfile]: Ok. I'll try to remove it from this branch.

File: lib/api.rb (hunk 2 above)
Comment (Pschittt) [reply lib/api.rb]: Ok.

File: lib/api.rb (hunk 4 above)
Comment (Pschittt) [reply lib/api.rb]: Yes. If someone try to "guess" the username or password, we shouldn't say which one is wrong.

File: lib/api.rb (hunk 20 above)
Comment (Pschittt) [reply lib/api.rb]: Currently, nothing. But it needs to be implemented to do the "mass import". Like importing from 1password, lastpass, or Keepass.
The web vault manage to parse all the formats, but it send the parameters via POST command. I couldn't determine the link between items and folders. So i didn't implemented it.

File: lib/api.rb (hunk 23 above)
Comment (Pschittt) [reply lib/api.rb]: Just to "remember", or if someone want to implement it.

File: lib/api.rb (hunk 9 above)
Comment (Pschittt) [reply lib/api.rb]: Ok for the privatekey. But for the public key, hard to find the regex.

**Comment by Pschittt** [comment]
@jcs I've updated the api.rb with your advice.
One thing I didn't do, is the publickey check. If I did, i would have done a simple regex as (.*). I don't think it's really relevant this kind of regex.

One more thing, the parameter are all put downcase, except the one for publickey and encryptedkey during registering. They are passed as : [:keys][publicKey] and [:keys][encryptedPrivateKey]

Last question, concerning the library you are using for sqlite. Does this library prevent from SQL injection attacks ? Does the library check if the parameters sent, don't hide another SQL request ?

File: lib/api.rb (hunk 5 above)
Comment (andrzejwp) [reply lib/api.rb]: @Pschittt I think it's expected that you reply to this, somehow ;)

Comment (Pschittt) [reply lib/api.rb]: @andrzejwp
I’ve updated the code to Check that devicename and devicetype are present.
:)

**Comment by Pschittt** [comment]
@jcs Is it better now ? Can you merge it ?

File: lib/api.rb (hunk 4 above)
Comment (andrzejwp) [reply lib/api.rb]: @Pschittt I think the problem is in the wording - you wrote "Invalid password or password", while you meant "Invalid username or password". Can you please amend?

Comment (Pschittt) [reply lib/api.rb]: Indeed. Forgot to change it. Done with the last commit.

**Comment by andrzejwp** [comment]
@jcs can you please review the changes @Pschittt added and possibly merge it?

**Comment by logic** [comment]
Only one small style comment, without having spent a lot of time looking through the diff yet: there's a fair bit of mixed tab/space and trailing whitespace in your diff that you might want to clean up to match the style of the rest of the project.

File: lib/api.rb [hunk 24]
```diff
@@ -349,6 +497,51 @@ def validation_error(msg)
...
+    end
+	  {
+	    "Data" => d.user.folders.map{|f| f.to_hash},
+	    "Object" => "list",
+	  }.to_json
+  end
+  
+  get "/collections" do
+	  response['access-control-allow-origin'] = '*'
+	  {"Data":[],"Object":"list"}.to_json
```
Comment (logic) [inline lib/api.rb]: This should probably be `{"Data" => [], "Object" => "list"}.to_json`.

**Comment by Pschittt** [comment]
@logic I've done the update. I saw only 2 lines which the tab wheren't correct.

I also made the change you requested.

Hope it's ok now ;)

**Comment by logic** [comment]
FWIW, I see 85 lines with leading tabs after applying your diff; you might want to check your editor's settings, I suspect it's hiding them from you. `grep -P '^\t' api.rb` can show you the lines that have problems, `sed -ie 's/^\t/  /' api.rb` will get rid of them.

**Comment by Pschittt** [comment]
@logic Thanks !
Ran you sed command, and seems to did the trick.

I also added a OPTIONS request for HTTP CORS checks, which sometimes occurs.

Hope now everything is ok, so we can go forward with your other PR ;)

**Comment by Pschittt** [comment]
Hello,
Just added the import tool feature from the web vault.
Tried with keepassx (csv) only. If someone want to try with other type.

I think the code can be optimized by someone familiar with ruby.

@jcs Any plan on merging this pull request one day ?

File: lib/api.rb (hunk 5 above)
Comment (jcs) [reply lib/api.rb]: I'm still not seeing where you're enforcing that these parameters are there.

Comment (jcs) [reply lib/api.rb]: I'm still not seeing where you're enforcing that these parameters are there.

File: lib/api.rb [hunk 25]
```diff
@@ -243,59 +314,159 @@ def validation_error(msg)
...
+      return validation_error("Invalid key")
     end
 
-    begin
-      Bitwarden::CipherString.parse(params[:name])
-    rescue Bitwarden::InvalidCipherString
-      return validation_error("Invalid name")
+    d.user.private_key = params[:encryptedprivatekey]
+    d.user.public_key = params[:publickey]
+
```
Comment (jcs) [inline lib/api.rb]: Don't you need to save the user record to retain the private key and public key?

File: lib/api.rb [hunk 26]
```diff
@@ -243,59 +314,159 @@ def validation_error(msg)
...
-        "Edit" => true,
-      }).to_json
-    end
+   begin
+     Bitwarden::CipherString.parse(params[:key])
+   rescue Bitwarden::InvalidCipherString
+     return validation_error("Invalid key")
+   end
+
+   if d.user.password_hash == params[:masterpasswordhash]
```
Comment (jcs) [inline lib/api.rb]: You need to use `User#has_password_hash?` which is timing safe.

File: lib/api.rb [hunk 27]
```diff
@@ -243,59 +314,159 @@ def validation_error(msg)
...
+      c = Cipher.new
+      c.user_uuid = d.user_uuid
+      c.update_from_params(p)
+      c.folder_uuid = Folder.find_by_user_uuid_and_name(d.user_uuid, params[:folders][params[:folderrelationships][i]["value"].to_i]["name"]).uuid
+      Cipher.transaction do
+        if !c.save
+          return validation_error("error saving")
+        end
+      end
+      i+=1
```
Comment (jcs) [inline lib/api.rb]: This should just use `params[:ciphers].each_with_index do |p,i|` and avoid having to initialize `i` and increment it manually.

File: lib/api.rb (hunk 5 above)
Comment (Pschittt) [reply lib/api.rb]: if statement on line 176, line 190, line 194

File: lib/api.rb (hunk 25 above)
Comment (Pschittt) [reply lib/api.rb]: It's done like the others parameters just above, and checked with the if statement line 288.

File: lib/api.rb (hunk 5 above)
Comment (jcs) [reply lib/api.rb]: That doesn't enforce that the parameters are there, that just conditionally uses them if they are.  When they were in  `need_params`, the processing would stop and return an error if the parameters weren't there.

Comment (Pschittt) [reply lib/api.rb]: But if you enforce these 2 parameters, that breaks the web vault compatibility, as the web vault doesn't send them.

We can only check if they are here to use them.

**Comment by Pschittt** [comment]
Updated with last @jcs review.

**Comment by universal** [comment]
hmm.. I was just scrolling through the changes and I thought, that it should be possible to discern if the request is from the web-vault or not, and then one could basically split the execution path based on that, where necessary. This might make some cases when params are required or not clearer. Ignore this comment if it is nonsense.

Another thing i noticed is that in ruby you can use ```unless some_condition``` instead of ```if !some_condition``` .

And another thing I asked myself is and haven't investigated yet: Can we provide the web-vault code and serve that either by the proxy server in front or by the sinatra app itself ( i think there is a static/public files plugin/support). Then one probably also wouldn't need the CORS stuff :-)

**Comment by Pschittt** [comment]
I think that my piece of code cannot be more basic, so if it can be optimized or improved, it would be great.

About proxy, that’s what I did. I put a caddy proxy in front of the web vault and the API. And it is the proxy which redirect the request to the web vault angularjs app or to the API backend.

**Comment by universal** [comment]
Cool. Some explanation in the readme about getting/compiling the web-vault might be nice :-) I might have some time on the weekend to have a closer look at the code in this pull-request :-)

**Comment by Pschittt** [comment]
Everything is explained on the official bitwarden github to setup the web vault :
https://github.com/bitwarden/web/blob/master/README.md

File: lib/api.rb [hunk 28]
```diff
@@ -59,6 +100,12 @@ def validation_error(msg)
     if !js.strip.blank?
       params.merge!(JSON.parse(js))
     end
+  ## needed for the web vault, which doesn't use the content-type
+  elsif request.accept.to_s.match(/application\/json/) && !request.content_type.to_s.match(/application\/x-www-form-urlencoded/)
+    js = request.body.read.to_s
+    if !js.strip.blank?
+      params.merge!(JSON.parse(js))
```
Comment (universal) [inline lib/api.rb]: i don't think this is actually needed. Even though I haven't done an extensive trial of all options in the webvault, but adding and importing of keys does not need this.

File: lib/api.rb [hunk 29]
```diff
@@ -69,13 +116,23 @@ def validation_error(msg)
 
   # we're always going to reply with json
   content_type :json
+  response.headers['Access-Control-Allow-Origin'] = '*'
```
Comment (universal) [inline lib/api.rb]: if the web-app is hosted under the same domain as the backend, then this is also not needed.

File: lib/api.rb [hunk 30]
```diff
@@ -69,13 +116,23 @@ def validation_error(msg)
...
   #  refresh_token: just generate a new access_token
   # respond with an access_token and refresh_token
+  
+  # For HTTP CORS verification
+  options "*" do
+    response.headers["Allow"] = "GET, POST, OPTIONS, PUT, DELETE"
+    response.headers["Access-Control-Allow-Headers"] = "Authorization, Content-Type, Accept"
+    response.headers["Access-Control-Allow-Origin"] = "*"
+    200
+  end
```
Comment (universal) [inline lib/api.rb]: if the web-app is hosted under the same domain as the backend, then this is also not needed.

File: lib/api.rb (hunk 5 above)
Comment (universal) [reply lib/api.rb]: One can distinguish between web-vault requests and app-requests, and then require those two parameters or not. This should be possible with something like this:

```
    def web_vault_request?
      !!(params[:client_id].to_s == "web")
    end
```

File: lib/api.rb [hunk 31]
```diff
@@ -157,19 +218,31 @@ def validation_error(msg)
         return validation_error("Unknown error")
       end
 
+      headers "access-control-allow-origin" => "*"
```
Comment (universal) [inline lib/api.rb]: replication from the before filter, so can be removed.

File: lib/api.rb [hunk 32]
```diff
@@ -157,19 +218,31 @@ def validation_error(msg)
...
+
       {
         :access_token => d.access_token,
         :expires_in => (d.token_expires_at - Time.now).floor,
         :token_type => "Bearer",
         :refresh_token => d.refresh_token,
         :Key => d.user.key,
+        :PrivateKey => d.user.private_key,
         # TODO: when to include :privateKey and :TwoFactorToken?
       }.to_json
```
Comment (universal) [inline lib/api.rb]: add the PrivateKey to user.to_hash and then just use this instead of writing out the hash manually.

File: lib/api.rb [hunk 33]
```diff
@@ -157,19 +218,31 @@ def validation_error(msg)
...
 namespace BASE_URL do
+
+  # For HTTP CORS verification
+  options "*" do
+    response.headers["Allow"] = "GET, POST, OPTIONS, PUT, DELETE"
+    response.headers["Access-Control-Allow-Headers"] = "Authorization, Content-Type, Accept"
+    response.headers["Access-Control-Allow-Origin"] = "*"
+    200
+  end
+
```
Comment (universal) [inline lib/api.rb]: not needed if web-vault is hosted under the same domain as the backend, which I guess would be the usual setup.

File: lib/api.rb [hunk 34]
```diff
@@ -219,6 +298,7 @@ def validation_error(msg)
         return validation_error("User save failed")
       end
 
+      headers "access-control-allow-origin" => "*"
```
Comment (universal) [inline lib/api.rb]: repeated from the before filter

File: lib/api.rb (hunk 25 above)
Comment (universal) [reply lib/api.rb]: do we need to save the private and public key to the database?

File: lib/api.rb [hunk 35]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+      "Premium" => d.user.premium,
+      "MasterPasswordHint" => d.user.password_hint,
+      "Culture" => d.user.culture,
+      "TwoFactorEnabled" => d.user.totp_secret,
+      "Key" => d.user.key,
+      "PrivateKey" => d.user.private_key,
+      "SecurityStamp" => d.user.security_stamp,
+      "Organizations" => "[]",
+      "Object" => "profile",
+   }.to_json
```
Comment (universal) [inline lib/api.rb]: again: i think it should be possible to adopt the ```user.to_hash``` function and then use it instead of the manual hash.

File: lib/api.rb [hunk 36]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+     "Premium" => d.user.premium,
+     "MasterPasswordHint" => d.user.password_hint,
+     "Culture" => d.user.culture,
+     "TwoFactorEnabled" => d.user.totp_secret,
+     "Key" => d.user.key,
+     "PrivateKey" => d.user.private_key,
+     "SecurityStamp" => d.user.security_stamp,
+     "Organizations" => "[]",
+     "Object" => "profile",
+    }.to_json
```
Comment (universal) [inline lib/api.rb]: again: i think it should be possible to adopt the ```user.to_hash``` function and then use it instead of the manual hash.

File: lib/api.rb [hunk 37]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+   end
+   ""
   end
 
-  # update a cipher
-  put "/ciphers/:uuid" do
+  # Used to update email
+  post "/accounts/email-token" do
+   content_type :json
+   response['access-control-allow-origin'] = '*'
```
Comment (universal) [inline lib/api.rb]: repeated from the before filter

File: lib/api.rb [hunk 38]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+  end
+
+  #
+  # ciphers
+  #
+
+  # Import from keepass or others via web vault
+  post "/ciphers/import" do
+    content_type :json
+    response['access-control-allow-origin'] = '*'
```
Comment (universal) [inline lib/api.rb]: repeated from the before filter

File: lib/api.rb [hunk 39]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+        end
+      end
+    end
+
+    # We create each CipherString
+    params[:ciphers].each_with_index do |p,i|
+      c = Cipher.new
+      c.user_uuid = d.user_uuid
+      c.update_from_params(p)
+      c.folder_uuid = Folder.find_by_user_uuid_and_name(d.user_uuid, params[:folders][params[:folderrelationships][i]["value"].to_i]["name"]).uuid
```
Comment (universal) [inline lib/api.rb]: this fails if the cipher does not belong to a folder, i replaced it with the following:
```
find_folder_uuid = lambda do |i, user_uuid:|
  folder_index = params.dig(:folderrelationships, i, "value")
  if folder_index
    folder_name = params.dig(:folders, folder_index.to_i, "name")
    Folder.find_by_user_uuid_and_name(d.user_uuid, folder_name).uuid
  else
    nil
  end
end
c.folder_uuid = find_folder_uuid.call(i, user_uuid: d.user_uuid)
```

File: lib/api.rb [hunk 40]
```diff
@@ -243,59 +323,157 @@ def validation_error(msg)
...
+          return validation_error("error saving")
+        end
+      end
+    end
+    ""
+  end
+
+  # create a new cipher
+  post "/ciphers" do
+    response['access-control-allow-origin'] = '*'
```
Comment (universal) [inline lib/api.rb]: again: repeated from the before filter, so can be removed

File: lib/api.rb [hunk 41]
```diff
@@ -349,6 +540,51 @@ def validation_error(msg)
   # folders
   #
 
+  # retrieve folder
+  get "/folders" do
+    content_type :json
+    response['access-control-allow-origin'] = '*'
```
Comment (universal) [inline lib/api.rb]: again: repeated from the before filter and can be removed

**Comment by Pschittt** [comment]
@universal I think that the web vault evolved since I started this PR.

Like you said, it may be possible that some piece of code is not needed anymore.

What I can say, it’s that it works with my web vault version which is from end of november. If I have time, i will try to test it with an updated version.

File: lib/api.rb (hunk 28 above)
Comment (Pschittt) [reply lib/api.rb]: It’s possible that it is not needed anymore

File: lib/api.rb (hunk 29 above)
Comment (Pschittt) [reply lib/api.rb]: You’re right. But it need a reverse proxy in front (that’s what i did).

File: lib/api.rb (hunk 30 above)
Comment (Pschittt) [reply lib/api.rb]: You’re right. But it need a reverse proxy in front (that’s what i did).

File: lib/api.rb (hunk 33 above)
Comment (Pschittt) [reply lib/api.rb]: You’re right. But it need a reverse proxy in front (that’s what i did).

**Comment by universal** [comment]
@Pschittt hmm... If i find the time I will try to submit a pull request, but this might take some time. In the meantime you can probably use my additions branch (https://github.com/universal/bitwarden-ruby/tree/additions) and merge your deploy config in. It also has attachments support from both the web-vault and the clients.

jcs is most likely busy with other stuff and as this is only a side-project I can understand the low priority :-)

**Comment by mprasil** [comment]
If anybody want's to give @universal's additions branch a try, I've created a Docker image with Web and API packed in. https://hub.docker.com/r/mprasil/bitwarden-ruby/

```
docker run -d -v /bitwarden_data/:/bitwarden/api/db/ -p 8080:8080 mprasil/bitwarden-ruby
```

In case you want to build your own image, build code can be fetched from my repo here: https://github.com/mprasil/docker-bitwarden

**Comment by Pschittt** [comment]
Sorry. Don't have time to continue this.

BTW @universal seems to be more advanced on this now.
